Adds shared code between Orders and Cart calculation logic.

vedanshujain opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #24695 #24633 #24540 . This adds a trait which contains shared logic in orders and carts in one place.

### How to test the changes in this Pull Request:

There are multiple things to test in this pull request: 

For issue #24695 

1. Set tax to 7%, price inclusive of taxes and round at subtotal.
2. Add 2 products for $2 and $2.5
3. Complete an order with 1 quantity of $2 product and 4 quantities of $2.5 product. In cart the total should be $12. 
4. Go to Edit order from admin view.  Click on recalculate. Without this patch, order total will change to $12.01. With this patch it will stay at $12.00.

For issue #24633

1. Set tax to 20%, enabled on shipping as well, price inclusive of taxes and round at subtotal.
2. Add a coupon for $5 flat cart discount.
3. Add a product for $8.99.
4. Add shipping for $7.99 flat rate.
5. Add the product (step 3), and apply the coupon (step 2) to the cart. Complete the order with Shipping selected (after-tax it should be 9.59). The total in the cart should be 13.58
6. Go to the admin view for the order. Set the status to payment-pending and click on recalculate. Without this patch, the discount for the coupon will change to $4.16, but with this patch, the discount will stay at $4.17.

For issue #24540 

1. Set tax to 20%, price inclusive of taxes, and round at subtotal.
2. Add products for $498, $99, $248, $44 and $69.
3. Add all these products in the cart and complete the order.
4. In order received page, without this patch subtotal will be 957.99, but with this patch, it will be correctly set to 958.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

Dev - Refactor to use the same rounding logic in Order and Cart.

vedanshujain committed:
Add trait which contains common total calc logic.

`WC_Abstract_Order` and `WC_Cart_Totals` have their own logic to calculate totals. This means that we would have to fix in two places. This commit adds a trait which can be used to place shared logic between above two classes.

vedanshujain committed:
Refactor Order Class to use shared calculation logic.

We added a trait to move shared logic betweem Orders and Cart. This commit refactors Order class to use that shared logic.
Also adds a test for a failing case.

kloon commented on includes/abstracts/abstract-wc-order.php:
```diff
@@ -769,6 +770,23 @@ public function get_items( $types = 'line_item' ) {
+	 *
+	 * @return array Array of values.
+	 */
+	protected function get_values_for_total( $field ) {
```
Perhaps a more general helper function would be better instead of having to duplicate this function here and in WC_Cart_Totals. I like the use of traits but it seems it defeats the purpose if we are overriding it with custom versions anyway in the different classes. There should be a way to make the trait function support both scenarios if we pass in the item/s

I am trying to avoid us having code in different places that are trying to achieve the same end goal, especially when it comes to taxes and calculations having duplicate functions is what is causing the current discrepancies we see.

vedanshujain replied:
Refactored Trait class to don't this abstract method. This is a bit different from what you suggested but now Trait only contains functionality which is explicitly related to calculating totals and nothing else. I did not move WC_Order and WC_Cart specific logic inside the trait because then it was not looking cleaner, and we may end up using this logic in some other place as well, so probably better to keep specific code in that class. Wdyt?

kloon reviewed (commented):
Just got a suggestion/question, let me know what you think.

vedanshujain committed:
Refactor so that trait don't have class specific functions.

vedanshujain committed:
Use `wc_round_tax_total` to make it consistent with how we round taxes.

When price includes taxes, we round down. This patch uses the correct method to round taxes depending on the mode instead of plain rounding.

vedanshujain committed:
Use standard way to calculate total in `get_subtotal_to_display` for consistency.

We were calculating subtotal to display slightly differently then we would have calculated subtotal in the cart. This was affecting subtotal value in invoices and in order confirmation screen.

This patch updates how we calculate subtotal to display such that we calculate in same way.

vedanshujain commented:
Update: Edited description to include more testing instructions.

claudiosanches reviewed (approved):
Great work!

kevinvettorel commented:
Hello, I do not speak English very well so I wanted to be sure to understand ... What files should I change to no longer have this problem of penny? These files : https://github.com/woocommerce/woocommerce/pull/24828#issuecomment-558132587

I had this problem, if I do the same manipulation as you, everything should be ok?
https://github.com/woocommerce/woocommerce/issues/24540#issuecomment-544500774

Thanks in advance for your help! Kevin

vedanshujain commented:
@kevinvettorel This patch will be released with WooCommerce 3.9 in early January 2020. I would recommend you to wait till then and update when it's available.

gregBerthelot commented:
Seems that this issue is back in 5.8.0 as we can see here https://cln.sh/ZZJ2Ap 

Product tax excluded : 18,96
Tax : 5,5%
Product tax included = 18,96 + 18,96 * 5,5 / 100 = 20,0028 displayed as 20,00 for 1 product
2 Products tax included = 2 * (18,96 + 18,96 * 5,5 / 100) = 40,0056 displayed as 40,01 for 1 product
